# Introduction

In order to explain what have been obtained, this report will be used. First of all, general information related to dataset can be listed as:

## Pisa Reading Score
* Target Variable is Reading Score.
* Both Train and Test dataset have more than 200 input.
* Dataset has more than 20 Features.
* A brief description added to the notebooks at the beggining. 

## Blog Comment
* Target Variable is Reading Score.
* Both Train and Test dataset have more than 200 input.
* Dataset has more than 20 Features.
* A brief description added to the notebooks at the beggining. 

## Student Evaluation
* Target Variable is Reading Score.
* Both Train and Test dataset have more than 200 input.
* Dataset has more than 20 Features.
* A brief description added to the notebooks at the beggining. 

## Spam Mail
* Target Variable is Reading Score.
* Both Train and Test dataset have more than 200 input.
* Dataset has more than 20 Features.
* A brief description added to the notebooks at the beggining. 

## Specific Requirements
In addition to this requirements, **specific information** related to dataset can be listed as:

* Pisa Reading Score and Blog Comment Dataset have regression Problem.
* Student Evaluation Test is multi-classification problem.
* Spam Mail has imbalance problem in the classification problem.
* Blog Comment and Spam Mail datasets have more than 50 features.
* Student Evaluation, Reading Score Datasets have categorical independent variables.

# Achieved Performance Measures for Datasets

First of all, what obtained in the datasets are added to this report. After these performance results, general comments will be made related to models and datasets. Namely, specific information related to models and dataset relationship can be found in the comment in the performance measure parts. At the end, a brief comparision and models' performance in different type of problems will be evaluated.

# Pisa Reading Score

## Linear Regression with Penalty

Different Penalty types are tried in the glmnet package by using mean absolute percentage error and mean square error objective. In these 2 penalty, 3 different lambdas are tried to obtain a model. In addition, all the possible lambda values are checked by plotting the cv.glmnet() results. Predictions are made based on these models. Best model performance are observed in this test dataset. After all, caret() packages are used to make a model can be comparable with other method's models. Namely, caret library is used in all the methods to obtain similar report types. Penalized Regression Approach is made with 3 different lambda values. These are lambda.min, lambda.1se, and random a lambda value achieved in the previous cv.glmnet() process. Model results are as follows:

In [70]:
lm_model

glmnet 

2414 samples
  23 predictor

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 2172, 2170, 2173, 2173, 2172, 2173, ... 
Resampling results across tuning parameters:

  lambda      RMSE      Rsquared   MAE     
   0.2935122  74.18294  0.3113987  59.11383
   4.7835176  75.29688  0.3020824  60.35636
  13.3104220  80.12992  0.2660190  64.68339

Tuning parameter 'alpha' was held constant at a value of 1
RMSE was used to select the optimal model using the smallest value.
The final values used for the model were alpha = 1 and lambda = 0.2935122.

### Best Model for Penalized Regression Approach

RMSE will be the objective parameter for evaluation of the model. In this perpective, **lambda 0.2935122(lambda.min) and MAE Penalty type** is the best tuned parameter for this model for Pisa Reading Score Dataset.

In [100]:
print(paste("The Smallest RMSE value in Penalized Regression Approach:",lm_model$results$RMSE[which.min(lm_model$results$RMSE)]))

[1] "The Smallest RMSE value in Penalized Regression Approach: 74.1829393463686"


## Decision Tree

**6 different minbucket** values are tried to get best model for Decision Tree. However, in caret library there is no available packages for rpart for tuning the minimal number of observation per tree leaf. So for loop is tried to get information related to effects in the change of number of variable in the tree leaf. In addition, all models tested with **10 different cp value**. After looking the created model's result, the best model is selected. The best RMSE value is selected with considering the minbucket values and cp values. Selected model evaluated 1 more time and named as tr_last. Model results are as follows:

In [92]:
tr_last

CART 

2414 samples
  29 predictor

No pre-processing
Resampling: Cross-Validated (10 fold) 
Summary of sample sizes: 2174, 2171, 2173, 2172, 2172, 2172, ... 
Resampling results across tuning parameters:

  cp    RMSE      Rsquared   MAE     
  0.00  80.54359  0.2271929  64.29627
  0.01  79.54121  0.2094880  63.67533
  0.02  81.09784  0.1780498  65.22498
  0.03  81.86471  0.1618356  65.90086
  0.04  81.86471  0.1618356  65.90086
  0.05  81.86471  0.1618356  65.90086
  0.06  83.80165  0.1216110  67.70118
  0.07  83.82850  0.1210650  67.73239
  0.08  83.82850  0.1210650  67.73239
  0.09  83.82850  0.1210650  67.73239
  0.10  83.82850  0.1210650  67.73239

RMSE was used to select the optimal model using the smallest value.
The final value used for the model was cp = 0.01.

### Best Model for Decision Tree

RMSE will be the objective parameter for evaluation of the model. In this perpective, **cp=0.01 and minbucket = 9** is the best tuned parameter for this model for Pisa Reading Score Dataset.

In [93]:
print(paste("The Smallest RMSE value in Decision Tree:",tr_last$results$RMSE[which.min(tr_last$results$RMSE)]))

[1] 79.54121

## Random Forest

As mentioned in the homework description, only mtry namely number of selected feature for model is modified. 4 different number is determined. Best model is achived the model with the biggest number of feature. Hovewer, Mean Absolute Error is the smallest in the model with the smallest number of feature. Model results are as follows:

In [49]:
rf_fit

Random Forest 

2414 samples
  23 predictor

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 2173, 2172, 2172, 2173, 2173, 2173, ... 
Resampling results across tuning parameters:

  mtry  RMSE      Rsquared   MAE     
   4    74.22778  0.3127169  59.68398
   8    74.86137  0.2988596  60.03758
  10    75.34474  0.2908063  60.46137
  15    75.50474  0.2895483  60.54514

Tuning parameter 'splitrule' was held constant at a value of variance

Tuning parameter 'min.node.size' was held constant at a value of 5
RMSE was used to select the optimal model using the smallest value.
The final values used for the model were mtry = 4, splitrule = variance
 and min.node.size = 5.

### Best Model for Random Forest

RMSE will be the objective parameter for evaluation of the model. In this perpective, **mtry=4** is the best tuned parameter for this model for Pisa Reading Score Dataset.

In [99]:
print(paste("The Smallest RMSE value in Random Forest:",rf_fit$results$RMSE[which.min(rf_fit$results$RMSE)]))

[1] "The Smallest RMSE value in Random Forest: 74.2277812566288"


## Stochastic Gradient Boosting

**Different 5 shrinkage value, 3 interaction.depth, 5 n.trees values** are tried to obtain the best model. This model is the one which takes the most time to create the models. The reason behind this fact caused from number of tuned parameter in the approach. However, it has the best RMSE value among the models. Model results are as follows:

In [50]:
gbm_fit

Stochastic Gradient Boosting 

2414 samples
  23 predictor

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 2173, 2172, 2172, 2171, 2173, 2174, ... 
Resampling results across tuning parameters:

  shrinkage  interaction.depth  n.trees  RMSE      Rsquared   MAE     
  0.1        1                   50      75.47460  0.3126352  60.62038
  0.1        1                  100      73.47254  0.3300000  58.69375
  0.1        1                  150      72.95822  0.3358354  58.19940
  0.1        1                  200      72.80733  0.3378907  58.00691
  0.1        1                  250      72.63771  0.3407722  57.90860
  0.1        3                   50      73.16656  0.3337547  58.63596
  0.1        3                  100      72.34571  0.3459466  57.79544
  0.1        3                  150      72.28070  0.3472914  57.76970
  0.1        3                  200      72.12923  0.3504537  57.65767
  0.1        3                  250      72

### Best Model for Stochastic Gradient Boosting

RMSE will be the objective parameter for evaluation of the model. In this perpective, **n.trees = 100 interaction.depth =
 5, shrinkage = 0.1 and n.minobsinnode = 20** is the best tuned parameter for this model for Pisa Reading Score Dataset.

In [97]:
print(paste("The Smallest RMSE value in Stochastic Gradient Boosting:",gbm_fit$results$RMSE[which.min(gbm_fit$results$RMSE)]))

[1] "The Smallest RMSE value in Stochastic Gradient Boosting: 72.1221273300718"


# Comparision of Results in Training and Test Dataset

In [126]:
train=list("Linear Regression with Lasso Results"=lm_model$results$RMSE[which.min(lm_model$results$RMSE)],
            "Decision Tree Results"=    tr_last$results$RMSE[which.min(tr_last$results$RMSE)],
            "Random Forest Results"=    rf_fit$results$RMSE[which.min(rf_fit$results$RMSE)],
            "Result Stochastic Gradient Boosting Results"=    gbm_fit$results$RMSE[which.min(gbm_fit$results$RMSE)])

## RMSE in The Train Dataset

In [129]:
t(train)

Linear Regression with Lasso Results Decision Tree Results
[1,] 74.18294                             79.54121             
     Random Forest Results Result Stochastic Gradient Boosting Results
[1,] 74.22778              72.12213

In [115]:
perf_dt("Pisa Data Set for Lasso Function with min lambda and mse objective", as.numeric(pisa_test$readingScore), prediction_pra_mse_pisa_min)
perf_dt("Decision Tree with CV for Pisa Dataset",pisa_test_dt$readingScore,as.numeric(predicted_pisa_dt))
perf_dt("First Data Set for Random Forest", as.numeric(RandomForest_pisa), as.numeric(pisa_test$readingScore))
perf_dt("First Data Set for Stochastic Gradient Boosting", as.numeric(predicted_pisa_sgb), as.numeric(pisa_test$readingScore))

name                                                               n  
1 Pisa Data Set for Lasso Function with min lambda and mse objective 990
  mean     sd       FBias       MAPE      RMSE     MAD      WMAPE    
1 519.8643 88.80043 0.007031004 0.1220864 2.435895 60.65549 0.1166756

name                                   n   mean     sd       FBias      
1 Decision Tree with CV for Pisa Dataset 990 519.8643 88.80043 0.007325851
  MAPE      RMSE     MAD      WMAPE    
1 0.1262667 2.506513 62.69629 0.1206013

name                             n   mean     sd       FBias        MAPE     
1 First Data Set for Random Forest 990 517.8192 45.48294 -0.003949342 0.1157326
  RMSE     MAD      WMAPE    
1 2.383545 59.41911 0.1147487

name                                            n   mean     sd      
1 First Data Set for Stochastic Gradient Boosting 990 516.5296 52.44005
  FBias       MAPE      RMSE     MAD     WMAPE    
1 -0.00645595 0.1154572 2.364434 58.9861 0.1141969

Models got similar performance as achieved in the training dataset. Namely, it is hard to say that there is an over-fitting or under-fitting. Reading Score Estimation can be described as regression problem, so having the worst performance measure both training and test data sets aren't a suprise for decision tree model. In addition to under/over-fitting, it can be said that error rates are consistent for both datasets, which can be observed looking the correlation between the results pf performance of the models in both dataset. 

# Blog Comment

## Linear Regression with Penalty

Different Penalty types are tried in the glmnet package by using mean absolute percentage error and mean square error objective. In these 2 penalty, 3 different lambdas are tried to obtain a model. In addition, all the possible lambda values are checked by plotting the cv.glmnet() results. Predictions are made based on these models. Best model performance are observed in this test dataset. After all, caret() packages are used to make a model can be comparable with other method's models. Namely, caret library is used in all the methods to obtain similar report types. Penalized Regression Approach is made with 3 different lambda values. These are lambda.min, lambda.1se, and random a lambda value achieved in the previous cv.glmnet() process. Model results are as follows:

In [43]:
lm_model

glmnet 

8023 samples
 280 predictor

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 7221, 7220, 7220, 7220, 7222, 7220, ... 
Resampling results across tuning parameters:

  lambda    RMSE      Rsquared   MAE      
  2.441102  24.42847  0.2050926   9.869664
  3.226993  24.61889  0.2009448  10.022668
  4.681814  25.07599  0.1863053  10.510031

Tuning parameter 'alpha' was held constant at a value of 1
RMSE was used to select the optimal model using the smallest value.
The final values used for the model were alpha = 1 and lambda = 2.441102.

### Best Model for Penalized Regression Approach

RMSE will be the objective parameter for evaluation of the model. In this perpective, **lambda 2.441102(lambda.min) and MAE Penalty type** is the best tuned parameter for this model for Blog Comment Dataset.

In [44]:
print(paste("The Smallest RMSE value in Penalized Regression Approach:",lm_model$results$RMSE[which.min(lm_model$results$RMSE)]))

[1] "The Smallest RMSE value in Penalized Regression Approach: 24.4284740530075"


## Decision Tree

**6 different minbucket** values are tried to get best model for Decision Tree. However, in caret library there is no available packages for rpart for tuning the minimal number of observation per tree leaf. So for loop is tried to get information related to effects in the change of number of variable in the tree leaf. In addition, all models tested with **10 different cp value**. After looking the created model's result, the best model is selected. The best RMSE value is selected with considering the minbucket values and cp values. Selected model evaluated 1 more time and named as tr_last. Model results are as follows:

In [45]:
tr_last

CART 

8023 samples
 280 predictor

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 7220, 7220, 7221, 7222, 7221, 7220, ... 
Resampling results across tuning parameters:

  cp    RMSE      Rsquared   MAE      
  0.00  22.57559  0.3897072   7.384471
  0.01  23.31100  0.2881720   8.693458
  0.02  23.59941  0.2619470   9.289978
  0.03  23.57051  0.2572654   9.434858
  0.04  24.22841  0.2119279   9.718694
  0.05  24.51101  0.1872444   9.984797
  0.06  24.51033  0.1872444   9.990837
  0.07  25.27038  0.1352021  10.349000
  0.08  25.27885  0.1329492  10.415419
  0.09  25.27885  0.1329492  10.415419
  0.10  25.27885  0.1329492  10.415419

RMSE was used to select the optimal model using the smallest value.
The final value used for the model was cp = 0.

### Best Model for Decision Tree

RMSE will be the objective parameter for evaluation of the model. In this perpective, **cp=0 and minbucket = 9** is the best tuned parameter for this model for Blog Comment Dataset.

In [46]:
print(paste("The Smallest RMSE value in Decision Tree:",tr_last$results$RMSE[which.min(tr_last$results$RMSE)]))

[1] "The Smallest RMSE value in Decision Tree: 22.5755890810299"


## Random Forest

As mentioned in the homework description, only mtry namely number of selected feature for model is modified. 4 different number is determined. Best model is achived the model with the biggest number of feature. Hovewer, Mean Absolute Error is the smallest in the model with the smallest number of feature. Model results are as follows:

In [47]:
rf_fit

Random Forest 

8023 samples
 280 predictor

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 7220, 7221, 7220, 7221, 7221, 7221, ... 
Resampling results across tuning parameters:

  mtry  RMSE      Rsquared   MAE     
   4    23.12233  0.3410507  9.550554
   8    20.40609  0.4818724  7.921203
  10    19.78936  0.5023333  7.523441
  15    19.20015  0.5178478  6.988483

Tuning parameter 'splitrule' was held constant at a value of variance

Tuning parameter 'min.node.size' was held constant at a value of 5
RMSE was used to select the optimal model using the smallest value.
The final values used for the model were mtry = 15, splitrule = variance
 and min.node.size = 5.

### Best Model for Random Forest

RMSE will be the objective parameter for evaluation of the model. In this perpective, **mtry=15** is the best tuned parameter for this model for Blog Comment Dataset.

In [48]:
print(paste("The Smallest RMSE value in Random Forest:",rf_fit$results$RMSE[which.min(rf_fit$results$RMSE)]))

[1] "The Smallest RMSE value in Random Forest: 19.2001454757668"


## Stochastic Gradient Boosting

**Different 5 shrinkage value, 3 interaction.depth, 5 n.trees values** are tried to obtain the best model. This model is the one which takes the most time to create the models. The reason behind this fact caused from number of tuned parameter in the approach. However, it has the best RMSE value among the models. Model results are as follows:

In [49]:
gbm_fit

Stochastic Gradient Boosting 

8023 samples
 270 predictor

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 7222, 7220, 7221, 7220, 7220, 7221, ... 
Resampling results across tuning parameters:

  shrinkage  interaction.depth  n.trees  RMSE      Rsquared   MAE     
  0.1        1                   25      23.51668  0.2849392  9.506556
  0.1        1                   50      22.80351  0.3032841  8.832547
  0.1        1                   75      22.65047  0.3098092  8.775545
  0.1        1                  100      22.63235  0.3104455  8.889400
  0.1        1                  125      22.61024  0.3117079  8.963804
  0.1        3                   25      21.73289  0.3743305  8.388996
  0.1        3                   50      21.29655  0.3899361  8.003750
  0.1        3                   75      21.22808  0.3937943  7.952887
  0.1        3                  100      21.24836  0.3933181  7.957694
  0.1        3                  125      21

### Best Model for Stochastic Gradient Boosting

RMSE will be the objective parameter for evaluation of the model. In this perpective, **n.trees = 125 interaction.depth =
 5, shrinkage = 0.1 and n.minobsinnode = 20** is the best tuned parameter for this model for Number of Blog Dataset.

In [50]:
print(paste("The Smallest RMSE value in Stochastic Gradient Boosting:",gbm_fit$results$RMSE[which.min(gbm_fit$results$RMSE)]))

[1] "The Smallest RMSE value in Stochastic Gradient Boosting: 20.7380054790217"


# Comparision of Results in Training and Test Dataset

In [51]:
train=list("Linear Regression with Lasso Results"=lm_model$results$RMSE[which.min(lm_model$results$RMSE)],
            "Decision Tree Results"=    tr_last$results$RMSE[which.min(tr_last$results$RMSE)],
            "Random Forest Results"=    rf_fit$results$RMSE[which.min(rf_fit$results$RMSE)],
            "Result Stochastic Gradient Boosting Results"=    gbm_fit$results$RMSE[which.min(gbm_fit$results$RMSE)])

## RMSE in The Train Dataset

In [52]:
t(train)

Linear Regression with Lasso Results Decision Tree Results
[1,] 24.42847                             22.57559             
     Random Forest Results Result Stochastic Gradient Boosting Results
[1,] 19.20015              20.73801

In [55]:
perf_dt("Blog Comment Data Set for Lasso Function with min lambda and mse objective", as.numeric(blog_test$target), prediction_pra_mse_blog_1se)
perf_dt("Decision Tree with CV for Blog Comment Dataset",blog_test$target,as.numeric(predicted_blog_dt))
perf_dt("First Data Set for Random Forest", as.numeric(blog_test$target), as.numeric(RandomForest_blog))
perf_dt("First Data Set for Stochastic Gradient Boosting", as.numeric(blog_test$target), as.numeric(predicted_blog_sgb) )

name                                                                      
1 Blog Comment Data Set for Lasso Function with min lambda and mse objective
  n    mean     sd       FBias      MAPE RMSE      MAD      WMAPE   
1 1977 7.152757 16.66787 -0.2301968 Inf  0.3298626 8.726155 1.219971

name                                           n    mean     sd      
1 Decision Tree with CV for Blog Comment Dataset 1977 7.152757 16.66787
  FBias     MAPE RMSE      MAD      WMAPE    
1 -0.211185 NaN  0.3807407 6.082319 0.8503462

name                             n    mean     sd       FBias      MAPE
1 First Data Set for Random Forest 1977 7.152757 16.66787 -0.2323384 Inf 
  RMSE      MAD      WMAPE    
1 0.2484009 5.514749 0.7709963

name                                            n    mean     sd      
1 First Data Set for Stochastic Gradient Boosting 1977 7.152757 16.66787
  FBias      MAPE RMSE      MAD      WMAPE    
1 -0.1954518 Inf  0.3035566 6.246727 0.8733313

Models got similar performance as achieved in the training dataset. Namely, it is hard to say that there is an over-fitting or under-fitting. Estimation of number of blog comment  can be described as regression problem, so having the worst performance measure both training and test data sets aren't a suprise for decision tree model. In addition to under/over-fitting, it can be said that error rates are consistent for both datasets, which can be observed looking the correlation between the results pf performance of the models in both dataset. 

# Student Evaluation

### Linear Regression with Penalty

In [54]:
lrp=confusionMatrix(data = as.factor(prediction_pra_mae_min), reference = as.factor(test$y), mode = "prec_recall")
lrp

Confusion Matrix and Statistics

          Reference
Prediction   1   2   3   4   5
         1 249  28  61  49  37
         2   0   1   0   1   0
         3  57  70 255 155  60
         4   9   8  23  29  14
         5   9   3  16  11  19

Overall Statistics
                                          
               Accuracy : 0.4751          
                 95% CI : (0.4461, 0.5042)
    No Information Rate : 0.305           
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.2724          
                                          
 Mcnemar's Test P-Value : < 2.2e-16       

Statistics by Class:

                     Class: 1  Class: 2 Class: 3 Class: 4 Class: 5
Precision              0.5873 0.5000000   0.4271  0.34940  0.32759
Recall                 0.7685 0.0090909   0.7183  0.11837  0.14615
F1                     0.6658 0.0178571   0.5357  0.17683  0.20213
Prevalence             0.2784 0.0945017   0.3050  0.21048  0.11

### Decision Tree

In [55]:
dtr=confusionMatrix(data = as.factor(predicted_std), reference = as.factor(test$y), mode = "prec_recall")
dtr

Confusion Matrix and Statistics

          Reference
Prediction   1   2   3   4   5
         1 257  14  40  24  32
         2   0   0   0   0   0
         3  67  96 315 221  98
         4   0   0   0   0   0
         5   0   0   0   0   0

Overall Statistics
                                          
               Accuracy : 0.4914          
                 95% CI : (0.4623, 0.5205)
    No Information Rate : 0.305           
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.277           
                                          
 Mcnemar's Test P-Value : NA              

Statistics by Class:

                     Class: 1 Class: 2 Class: 3 Class: 4 Class: 5
Precision              0.7003       NA   0.3952       NA       NA
Recall                 0.7932   0.0000   0.8873   0.0000   0.0000
F1                     0.7438       NA   0.5469       NA       NA
Prevalence             0.2784   0.0945   0.3050   0.2105   0.1117
D

### Random Forest

In [56]:
rfr=confusionMatrix(data = as.factor(RandomForest_std), reference = as.factor(test$y), mode = "prec_recall")
rfr

Confusion Matrix and Statistics

          Reference
Prediction   1   2   3   4   5
         1 242  11  37  20  23
         2   0  14   7   6   3
         3  58  70 246 137  56
         4  15  13  47  65  24
         5   9   2  18  17  24

Overall Statistics
                                          
               Accuracy : 0.5077          
                 95% CI : (0.4786, 0.5368)
    No Information Rate : 0.305           
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.3284          
                                          
 Mcnemar's Test P-Value : < 2.2e-16       

Statistics by Class:

                     Class: 1 Class: 2 Class: 3 Class: 4 Class: 5
Precision              0.7267  0.46667   0.4339  0.39634  0.34286
Recall                 0.7469  0.12727   0.6930  0.26531  0.18462
F1                     0.7367  0.20000   0.5336  0.31785  0.24000
Prevalence             0.2784  0.09450   0.3050  0.21048  0.11168
D

### Stochastic Gradient Boosting

In [57]:
sgbr=confusionMatrix(data = as.factor(predicted_sgb), reference = as.factor(test$y), mode = "prec_recall")
sgbr

Confusion Matrix and Statistics

          Reference
Prediction   1   2   3   4   5
         1 254  11  33  21  30
         2   1  13  12   4   2
         3  47  70 243 141  48
         4  15  12  45  65  26
         5   7   4  22  14  24

Overall Statistics
                                          
               Accuracy : 0.5146          
                 95% CI : (0.4855, 0.5437)
    No Information Rate : 0.305           
    P-Value [Acc > NIR] : < 2.2e-16       
                                          
                  Kappa : 0.3385          
                                          
 Mcnemar's Test P-Value : < 2.2e-16       

Statistics by Class:

                     Class: 1 Class: 2 Class: 3 Class: 4 Class: 5
Precision              0.7278  0.40625   0.4426  0.39877  0.33803
Recall                 0.7840  0.11818   0.6845  0.26531  0.18462
F1                     0.7548  0.18310   0.5376  0.31863  0.23881
Prevalence             0.2784  0.09450   0.3050  0.21048  0.11168
D

By looking the Accuracy, it can be said that best performance are optained in the Random Forest and Stochastic Gradient Boosting. Indeed, these 2 model's performance is too similar, which can be understood by looking the Precision, Recall, and F1 score of the models. The biggest difference occurs in the precision and recall values of the model. So, main objective can determine which model must be used. Moreover, performance of the other two model is also similar. Indeed decision tree's approach is more proper for classification problems. The reason behind the similarity can be explained by complexity of the models. Namely, Decision Tree's complexity is less than constructed Linear Regression with lasso, but it still has similar performance compared to the other.

# Conclusion of Models

## Linear Regression with Penalty

Different Penalty types are tried in the glmnet package by using mean absolute percentage error and mean square error objective. In these 2 penalty, 3 different lambdas are tried to obtain a model. In addition, all the possible lambda values are checked by plotting the cv.glmnet() results. Predictions are made based on these models. Best model performance are observed in this test dataset. After all, caret() packages are used to make a model can be comparable with other method's models. Namely, caret library is used in all the methods to obtain similar report types. Penalized Regression Approach is made with 3 different lambda values. These are lambda.min, lambda.1se, and random a lambda value achieved in the previous cv.glmnet() process. Model results are as follows:

In [58]:
lm_model

glmnet 

4656 samples
  32 predictor
   5 classes: '1', '2', '3', '4', '5' 

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 4190, 4190, 4190, 4191, 4190, 4192, ... 
Resampling results across tuning parameters:

  lambda       Accuracy   Kappa    
  0.002179314  0.5199788  0.3445326
  0.007304193  0.5158988  0.3296490
  0.098829223  0.3047682  0.0000000

Tuning parameter 'alpha' was held constant at a value of 1
Accuracy was used to select the optimal model using the largest value.
The final values used for the model were alpha = 1 and lambda = 0.002179314.

### Best Model for Penalized Regression Approach

RMSE will be the objective parameter for evaluation of the model. In this perpective, **alpha = 1, lambda = 0.002179314 and MAE Penalty type** is the best tuned parameter for this model for Student Evaluation Dataset.

In [59]:
print(lrp$overall[1])

 Accuracy 
0.4750859 


## Decision Tree

**6 different minbucket** values are tried to get best model for Decision Tree. However, in caret library there is no available packages for rpart for tuning the minimal number of observation per tree leaf. So for loop is tried to get information related to effects in the change of number of variable in the tree leaf. In addition, all models tested with **10 different cp value**. After looking the created model's result, the best model is selected. The best Accuracy value is selected with considering the minbucket values and cp values. Selected model evaluated 1 more time and named as tr_last. Model results are as follows:

In [75]:
tr_last

CART 

4656 samples
 164 predictor
   5 classes: '1', '2', '3', '4', '5' 

No pre-processing
Resampling: Cross-Validated (10 fold) 
Summary of sample sizes: 4190, 4192, 4190, 4190, 4191, 4191, ... 
Resampling results across tuning parameters:

  cp    Accuracy   Kappa    
  0.00  0.4641210  0.2867510
  0.01  0.4935627  0.2805461
  0.02  0.4935627  0.2805461
  0.03  0.4935627  0.2805461
  0.04  0.4935627  0.2805461
  0.05  0.4935627  0.2805461
  0.06  0.4935627  0.2805461
  0.07  0.4935627  0.2805461
  0.08  0.4935627  0.2805461
  0.09  0.4935627  0.2805461
  0.10  0.4935627  0.2805461

Accuracy was used to select the optimal model using the largest value.
The final value used for the model was cp = 0.1.

### Best Model for Decision Tree

RMSE will be the objective parameter for evaluation of the model. In this perpective, **cp=0.1 and minbucket = 8** is the best tuned parameter for this model for Student Evaluation Dataset.

In [61]:
print(dtr$overall[1])

 Accuracy 
0.4914089 


## Random Forest

As mentioned in the homework description, only mtry namely number of selected feature for model is modified. 4 different number is determined. Best model is achived the model with the 10 number of feature. Model results are as follows:

In [62]:
rf_fit

Random Forest 

4656 samples
  32 predictor
   5 classes: '1', '2', '3', '4', '5' 

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 4190, 4191, 4190, 4190, 4190, 4192, ... 
Resampling results across tuning parameters:

  mtry  Accuracy   Kappa    
   4    0.4374783  0.2109921
   8    0.4819396  0.2853275
  10    0.4886044  0.2969695
  15    0.4920430  0.3057549

Tuning parameter 'splitrule' was held constant at a value of extratrees

Tuning parameter 'min.node.size' was held constant at a value of 5
Accuracy was used to select the optimal model using the largest value.
The final values used for the model were mtry = 15, splitrule = extratrees
 and min.node.size = 5.

### Best Model for Random Forest

RMSE will be the objective parameter for evaluation of the model. In this perpective, **mtry=15** is the best tuned parameter for this model for Student Evaluation Dataset.

In [63]:
print(rfr$overall[1])

Accuracy 
0.507732 


## Stochastic Gradient Boosting

**Different 5 shrinkage value, 3 interaction.depth, 5 n.trees values** are tried to obtain the best model. This model is the one which takes the most time to create the models. The reason behind this fact caused from number of tuned parameter in the approach. However, it has the best Accuracy value with Random Forest. Model results are as follows:

In [64]:
gbm_fit

Stochastic Gradient Boosting 

4656 samples
  32 predictor
   5 classes: '1', '2', '3', '4', '5' 

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 4191, 4191, 4191, 4190, 4190, 4189, ... 
Resampling results across tuning parameters:

  shrinkage  interaction.depth  n.trees  Accuracy   Kappa    
  0.1        1                   50      0.5045147  0.3078019
  0.1        1                  100      0.5116004  0.3253585
  0.1        1                  150      0.5128862  0.3312142
  0.1        1                  200      0.5189012  0.3425569
  0.1        1                  250      0.5199714  0.3450251
  0.1        3                   50      0.5116027  0.3307548
  0.1        3                  100      0.5150413  0.3412033
  0.1        3                  150      0.5173949  0.3463261
  0.1        3                  200      0.5131017  0.3420885
  0.1        3                  250      0.5154663  0.3455269
  0.1        5                  

### Best Model for Stochastic Gradient Boosting

RMSE will be the objective parameter for evaluation of the model. In this perpective, **n.trees = 50 interaction.depth =
 5, shrinkage = 0.1 and n.minobsinnode = 20** is the best tuned parameter for this model for Student Evaluation Dataset.

In [65]:
print(sgbr$overall[1])

 Accuracy 
0.5146048 


# Comparision of Results in Training and Test Dataset

In [71]:
train=list("Linear Regression with Lasso Results"=lm_model$results$Accuracy[which.min(lm_model$results$Accuracy)],
            "Decision Tree Results"=    tr_last$results$Accuracy[which.min(tr_last$results$Accuracy)],
            "Random Forest Results"=    rf_fit$results$Accuracy[which.min(rf_fit$results$Accuracy)],
            "Result Stochastic Gradient Boosting Results"=    gbm_fit$results$Accuracy[which.min(gbm_fit$results$Accuracy)])

## Accuracy in The Train Dataset¶

In [72]:
t(train)

Linear Regression with Lasso Results Decision Tree Results
[1,] 0.3047682                            0.4652249            
     Random Forest Results Result Stochastic Gradient Boosting Results
[1,] 0.4374783             0.4525368

In [68]:
test=list("Linear Regression with Lasso Results"=lrp$overall[1],
            "Decision Tree Results"=    dtr$overall[1],
            "Random Forest Results"=    rfr$overall[1],
            "Result Stochastic Gradient Boosting Results"=    sgbr$overall[1])

## Accuracy in The Test Dataset

In [69]:
t(test)

Linear Regression with Lasso Results Decision Tree Results
[1,] 0.4750859                            0.4914089            
     Random Forest Results Result Stochastic Gradient Boosting Results
[1,] 0.507732              0.5146048

Models got similar performance as achieved in the training dataset. Namely, it is hard to say that there is an over-fitting or under-fitting. Student Evaluation values can be described as nulti-classification problem, so having the worst performance measure both training and test data sets aren't a suprise for Linear Regression model. In addition to under/over-fitting, it can be said that error rates are consistent for both datasets, which can be observed looking the correlation between the results pf performance of the models in both dataset. Interestingly, all the models have more accuracy in the test dataset. However, accuracy can be bad performance measure for this type dataset(imbalance in the dataset) because by having a bias in the major part of the class, model still can get better accuracy at the end. The reason behind this fact occured from the formula of the Accuracy measurement. So, F1-Score or Recall can be more reasonable to compare the models.

# Spam Mail

### Linear Regression with Penalty

In [70]:
lrp=confusionMatrix(data = as.factor(prediction_pra_mae_spam_min), reference = as.factor(test_spam$X1), mode = "prec_recall")
lrp

Confusion Matrix and Statistics

          Reference
Prediction   0   1
         0 538  33
         1  20 130
                                          
               Accuracy : 0.9265          
                 95% CI : (0.9049, 0.9445)
    No Information Rate : 0.7739          
    P-Value [Acc > NIR] : < 2e-16         
                                          
                  Kappa : 0.7838          
                                          
 Mcnemar's Test P-Value : 0.09929         
                                          
              Precision : 0.9422          
                 Recall : 0.9642          
                     F1 : 0.9531          
             Prevalence : 0.7739          
         Detection Rate : 0.7462          
   Detection Prevalence : 0.7920          
      Balanced Accuracy : 0.8809          
                                          
       'Positive' Class : 0               
                                          

### Decision Tree

In [71]:
dtr=confusionMatrix(data = as.factor(predicted_spam), reference = as.factor(test_spam$X1), mode = "prec_recall")
dtr

Confusion Matrix and Statistics

          Reference
Prediction   0   1
         0 533  30
         1  25 133
                                         
               Accuracy : 0.9237         
                 95% CI : (0.9019, 0.942)
    No Information Rate : 0.7739         
    P-Value [Acc > NIR] : <2e-16         
                                         
                  Kappa : 0.7796         
                                         
 Mcnemar's Test P-Value : 0.5896         
                                         
              Precision : 0.9467         
                 Recall : 0.9552         
                     F1 : 0.9509         
             Prevalence : 0.7739         
         Detection Rate : 0.7393         
   Detection Prevalence : 0.7809         
      Balanced Accuracy : 0.8856         
                                         
       'Positive' Class : 0              
                                         

### Random Forest

In [72]:
rfr=confusionMatrix(data = as.factor(PredictRandomForest_spam), reference = as.factor(test_spam$X1), mode = "prec_recall")
rfr

Confusion Matrix and Statistics

          Reference
Prediction   0   1
         0 547  18
         1  11 145
                                          
               Accuracy : 0.9598          
                 95% CI : (0.9427, 0.9729)
    No Information Rate : 0.7739          
    P-Value [Acc > NIR] : <2e-16          
                                          
                  Kappa : 0.8833          
                                          
 Mcnemar's Test P-Value : 0.2652          
                                          
              Precision : 0.9681          
                 Recall : 0.9803          
                     F1 : 0.9742          
             Prevalence : 0.7739          
         Detection Rate : 0.7587          
   Detection Prevalence : 0.7836          
      Balanced Accuracy : 0.9349          
                                          
       'Positive' Class : 0               
                                          

### Stochastic Gradient Boosting

In [73]:
sgbr=confusionMatrix(data = as.factor(predicted_spam_sgb), reference = as.factor(test_spam$X1), mode = "prec_recall")
sgbr

Confusion Matrix and Statistics

          Reference
Prediction   0   1
         0 545  16
         1  13 147
                                          
               Accuracy : 0.9598          
                 95% CI : (0.9427, 0.9729)
    No Information Rate : 0.7739          
    P-Value [Acc > NIR] : <2e-16          
                                          
                  Kappa : 0.8843          
                                          
 Mcnemar's Test P-Value : 0.7103          
                                          
              Precision : 0.9715          
                 Recall : 0.9767          
                     F1 : 0.9741          
             Prevalence : 0.7739          
         Detection Rate : 0.7559          
   Detection Prevalence : 0.7781          
      Balanced Accuracy : 0.9393          
                                          
       'Positive' Class : 0               
                                          

By looking the Accuracy, it can be said that best performance are optained in the Random Forest and Stochastic Gradient Boosting. Indeed, these 2 model's performance is too similar, which can be understood by looking the Precision, Recall, and F1 score of the models. The biggest difference occurs in the precision and recall values of the model. So, main objective can determine which model must be used. Moreover, performance of the other two model is also similar. Indeed decision tree's approach is more proper for classification problems. The reason behind the similarity can be explained by complexity of the models. Namely, Decision Tree's complexity is less than constructed Linear Regression with lasso, but it still has similar performance compared to the other.

# Conclusion of Models

## Linear Regression with Penalty

Different Penalty types are tried in the glmnet package by using mean absolute percentage error and mean square error objective. In these 2 penalty, 3 different lambdas are tried to obtain a model. In addition, all the possible lambda values are checked by plotting the cv.glmnet() results. Predictions are made based on these models. Best model performance are observed in this test dataset. After all, caret() packages are used to make a model can be comparable with other method's models. Namely, caret library is used in all the methods to obtain similar report types. Penalized Regression Approach is made with 3 different lambda values. These are lambda.min, lambda.1se, and random a lambda value achieved in the previous cv.glmnet() process. Model results are as follows:

In [74]:
lm_model

glmnet 

2880 samples
  57 predictor
   2 classes: '0', '1' 

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 2592, 2592, 2592, 2592, 2592, 2592, ... 
Resampling results across tuning parameters:

  lambda        Accuracy   Kappa    
  1.633465e-05  0.9309028  0.7962956
  1.050003e-04  0.9309028  0.7960720
  6.442722e-02  0.8225694  0.3270203

Tuning parameter 'alpha' was held constant at a value of 1
Accuracy was used to select the optimal model using the largest value.
The final values used for the model were alpha = 1 and lambda = 0.0001050003.

### Best Model for Penalized Regression Approach

RMSE will be the objective parameter for evaluation of the model. In this perpective, **alpha = 1, lambda = 0.0001050003 and MAE Penalty type** is the best tuned parameter for this model for Spam Mail Dataset.

In [82]:
print(lrp$overall[1])

Accuracy 
0.926491 


## Decision Tree

**6 different minbucket** values are tried to get best model for Decision Tree. However, in caret library there is no available packages for rpart for tuning the minimal number of observation per tree leaf. So for loop is tried to get information related to effects in the change of number of variable in the tree leaf. In addition, all models tested with **10 different cp value**. After looking the created model's result, the best model is selected. The best Accuracy value is selected with considering the minbucket values and cp values. Selected model evaluated 1 more time and named as tr_last. Model results are as follows:

In [86]:
tr_last

CART 

2880 samples
  57 predictor
   2 classes: '0', '1' 

No pre-processing
Resampling: Cross-Validated (10 fold) 
Summary of sample sizes: 2592, 2592, 2592, 2592, 2592, 2592, ... 
Resampling results across tuning parameters:

  cp    Accuracy   Kappa    
  0.00  0.9284722  0.7906848
  0.01  0.9246528  0.7753364
  0.02  0.9149306  0.7450010
  0.03  0.9086806  0.7235983
  0.04  0.9097222  0.7250175
  0.05  0.9097222  0.7250175
  0.06  0.9076389  0.7198955
  0.07  0.9076389  0.7198955
  0.08  0.9076389  0.7198955
  0.09  0.8979167  0.6961644
  0.10  0.8934028  0.6851602

Accuracy was used to select the optimal model using the largest value.
The final value used for the model was cp = 0.

### Best Model for Decision Tree

RMSE will be the objective parameter for evaluation of the model. In this perpective, **cp=0 and minbucket = 10** is the best tuned parameter for this model for Spam Mail Dataset.

In [87]:
print(dtr$overall[1])

 Accuracy 
0.9237171 


## Random Forest

As mentioned in the homework description, only mtry namely number of selected feature for model is modified. 4 different number is determined. Best model is achived the model with the 10 number of feature. Model results are as follows:

In [88]:
rf_fit

Random Forest 

2880 samples
  57 predictor
   2 classes: '0', '1' 

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 2592, 2592, 2592, 2592, 2592, 2592, ... 
Resampling results across tuning parameters:

  mtry  Accuracy   Kappa    
   4    0.9475694  0.8420506
   8    0.9538194  0.8625748
  10    0.9545139  0.8654099
  15    0.9527778  0.8597787

Tuning parameter 'splitrule' was held constant at a value of extratrees

Tuning parameter 'min.node.size' was held constant at a value of 5
Accuracy was used to select the optimal model using the largest value.
The final values used for the model were mtry = 10, splitrule = extratrees
 and min.node.size = 5.

### Best Model for Random Forest

RMSE will be the objective parameter for evaluation of the model. In this perpective, **mtry=10** is the best tuned parameter for this model for Spam Mail Dataset.

In [89]:
print(rfr$overall[1])

 Accuracy 
0.9597781 


## Stochastic Gradient Boosting

**Different 5 shrinkage value, 3 interaction.depth, 5 n.trees values** are tried to obtain the best model. This model is the one which takes the most time to create the models. The reason behind this fact caused from number of tuned parameter in the approach. However, it has the best Accuracy value with Random Forest. Model results are as follows:

In [90]:
gbm_fit

Stochastic Gradient Boosting 

2880 samples
  57 predictor
   2 classes: '0', '1' 

No pre-processing
Resampling: Cross-Validated (10 fold, repeated 1 times) 
Summary of sample sizes: 2592, 2592, 2592, 2592, 2592, 2592, ... 
Resampling results across tuning parameters:

  shrinkage  interaction.depth  n.trees  Accuracy   Kappa    
  0.1        1                   50      0.9267361  0.7729207
  0.1        1                  100      0.9416667  0.8245584
  0.1        1                  150      0.9447917  0.8349425
  0.1        1                  200      0.9454861  0.8373744
  0.1        1                  250      0.9451389  0.8367385
  0.1        3                   50      0.9454861  0.8371056
  0.1        3                  100      0.9500000  0.8522543
  0.1        3                  150      0.9513889  0.8565429
  0.1        3                  200      0.9517361  0.8579397
  0.1        3                  250      0.9552083  0.8685325
  0.1        5                   50      0.9493

### Best Model for Stochastic Gradient Boosting

RMSE will be the objective parameter for evaluation of the model. In this perpective, **n.trees = 250 interaction.depth =
 3, shrinkage = 0.1 and n.minobsinnode = 20** is the best tuned parameter for this model for Spam Mail Dataset.

In [92]:
print(sgbr$overall[1])

 Accuracy 
0.9597781 


# Comparision of Results in Training and Test Dataset

In [103]:
train=list("Linear Regression with Lasso Results"=lm_model$results$Accuracy[which.min(lm_model$results$Accuracy)],
            "Decision Tree Results"=    tr_last$results$Accuracy[which.min(tr_last$results$Accuracy)],
            "Random Forest Results"=    rf_fit$results$Accuracy[which.min(rf_fit$results$Accuracy)],
            "Result Stochastic Gradient Boosting Results"=    gbm_fit$results$Accuracy[which.min(gbm_fit$results$Accuracy)])

## Accuracy in The Train Dataset¶

In [104]:
t(train)

Linear Regression with Lasso Results Decision Tree Results
[1,] 0.8225694                            0.8934028            
     Random Forest Results Result Stochastic Gradient Boosting Results
[1,] 0.9475694             0.9267361

In [94]:
test=list("Linear Regression with Lasso Results"=lrp$overall[1],
            "Decision Tree Results"=    dtr$overall[1],
            "Random Forest Results"=    rfr$overall[1],
            "Result Stochastic Gradient Boosting Results"=    sgbr$overall[1])

## Accuracy in The Test Dataset

In [96]:
t(test)

Linear Regression with Lasso Results Decision Tree Results
[1,] 0.926491                             0.9237171            
     Random Forest Results Result Stochastic Gradient Boosting Results
[1,] 0.9597781             0.9597781

Models got similar performance as achieved in the training dataset. Namely, it is hard to say that there is an over-fitting or under-fitting. Spam Estimation can be described as classification problem, so having the worst performance measure both training and test data sets aren't a suprise for Linear Regression model. In addition to under/over-fitting, it can be said that error rates are consistent for both datasets, which can be observed looking the correlation between the results pf performance of the models in both dataset. However, there is still some problems related to consistency. It can be understood by looking the how much performance measure are changed between the datasets. Interestingly, all the models have more accuracy in the test dataset. However, accuracy can be bad performance measure for this type dataset(imbalance in the dataset) because by having a bias in the major part of the class, model still can get better accuracy at the end. The reason behind this fact occured from the formula of the Accuracy measurement. So, F1-Score or Recall can be more reasonable to compare the models.

# Final Thoughts

It can be said that trees' performance is better than linear regression in classification problems, which can be observed for both binary and multi classification. In addition, Random Forest and Stochastic Gradient Boosting' performance are better than normal decision tree. This fact exist not only in classification problem but also in regression problems. Morever, the worst performance observed in the multinomial classification. 

For the Regression problems, the worst performance belongs to Decision Tree. The reason behind this fact can be explained by the nature of the algorithms. Namely, extrapolation doesn't exist in the Decision Tree, which makes it performance to worse. In addition, Random Forest and Stochastic Gradient Boosting models' performance is better than the other tree model. And in some cases, these models have better performance compared to Linear Regression with Lasso. Indeed, this situation can be occured because of different reasons. However, the major difference can be complexity difference between the model. In other words, Random Forest and Stochastic Gradient Boosting have more complex stucture(not over-fitting). This situation appears by looking the how much times passed in the creation of the model. To sum up, Linear Regression model's nature is more proper for regression problems and the better model can be obtained in the limited time period by Linear Regression. 

In addition to performance measure, Stochastic Gradient Boosting has more flexibility to tune its parameters. This fact can cause extra effort to obtain the optimal parameter combination. However, tuning is a good option for getting the best model at the end. In addition, computers' CPU and GPU performance is more advanced and cloud systems exist in the world, which makes this model to the better among the other models. In addition, I have a project related to this XGBOOST library. In this project, I got a lot of evidence related to having good performance of XGBOOST. This fact is approved one more time in this homework.

Lastly, there is no problem related to overfitting. For some models(not too much complex) can have some underfitting problems in the training dataset. But still there is no solid evidence for this assumption. In addition, train error and test error of models have correlation. So, it can be said that all train and test datasets are consistent.